File: src/main/java/com/mssoftech/dbflute/bsentity/dbmeta/WhiteCompoundPkRefDbm.java
package com.mssoftech.dbflute.bsentity.dbmeta;

import java.util.List;
import java.util.Map;

import org.dbflute.Entity;
import org.dbflute.optional.OptionalEntity;
import org.dbflute.dbmeta.AbstractDBMeta;
import org.dbflute.dbmeta.info.*;
import org.dbflute.dbmeta.name.*;
import org.dbflute.dbmeta.property.PropertyGateway;
import org.dbflute.dbway.DBDef;
import com.mssoftech.dbflute.allcommon.*;
import com.mssoftech.dbflute.exentity.*;

/**
 * The DB meta of white_compound_pk_ref. (Singleton)
 * @author DBFlute(AutoGenerator)
 */
public class WhiteCompoundPkRefDbm extends AbstractDBMeta {

    // ===================================================================================
    //                                                                           Singleton
    //                                                                           =========
    private static final WhiteCompoundPkRefDbm _instance = new WhiteCompoundPkRefDbm();
    private WhiteCompoundPkRefDbm() {}
    public static WhiteCompoundPkRefDbm getInstance() { return _instance; }

    // ===================================================================================
    //                                                                       Current DBDef
    //                                                                       =============
    public String getProjectName() { return DBCurrent.getInstance().projectName(); }
    public String getProjectPrefix() { return DBCurrent.getInstance().projectPrefix(); }
    public String getGenerationGapBasePrefix() { return DBCurrent.getInstance().generationGapBasePrefix(); }
    public DBDef getCurrentDBDef() { return DBCurrent.getInstance().currentDBDef(); }

    // ===================================================================================
    //                                                                    Property Gateway
    //                                                                    ================
    // -----------------------------------------------------
    //                                       Column Property
    //                                       ---------------
    protected final Map<String, PropertyGateway> _epgMap = newHashMap();
    { xsetupEpg(); }
    protected void xsetupEpg() {
        setupEpg(_epgMap, et -> ((WhiteCompoundPkRef)et).getMultipleFirstId(), (et, vl) -> ((WhiteCompoundPkRef)et).setMultipleFirstId(cti(vl)), "multipleFirstId");
        setupEpg(_epgMap, et -> ((WhiteCompoundPkRef)et).getMultipleSecondId(), (et, vl) -> ((WhiteCompoundPkRef)et).setMultipleSecondId(cti(vl)), "multipleSecondId");
        setupEpg(_epgMap, et -> ((WhiteCompoundPkRef)et).getRefFirstId(), (et, vl) -> ((WhiteCompoundPkRef)et).setRefFirstId(cti(vl)), "refFirstId");
        setupEpg(_epgMap, et -> ((WhiteCompoundPkRef)et).getRefSecondId(), (et, vl) -> ((WhiteCompoundPkRef)et).setRefSecondId(cti(vl)), "refSecondId");
    }
    public PropertyGateway findPropertyGateway(String prop)
    { return doFindEpg(_epgMap, prop); }

    // -----------------------------------------------------
    //                                      Foreign Property
    //                                      ----------------
    protected final Map<String, PropertyGateway> _efpgMap = newHashMap();
    { xsetupEfpg(); }
    @SuppressWarnings("unchecked")
    protected void xsetupEfpg() {
        setupEfpg(_efpgMap, et -> ((WhiteCompoundPkRef)et).getWhiteCompoundPk(), (et, vl) -> ((WhiteCompoundPkRef)et).setWhiteCompoundPk((OptionalEntity<WhiteCompoundPk>)vl), "whiteCompoundPk");
    }
    public PropertyGateway findForeignPropertyGateway(String prop)
    { return doFindEfpg(_efpgMap, prop); }

    // ===================================================================================
    //                                                                          Table Info
    //                                                                          ==========
    protected final String _tableDbName = "white_compound_pk_ref";
    protected final String _tableDispName = "white_compound_pk_ref";
    protected final String _tablePropertyName = "whiteCompoundPkRef";
    protected final TableSqlName _tableSqlName = new TableSqlName("white_compound_pk_ref", _tableDbName);
    { _tableSqlName.xacceptFilter(DBFluteConfig.getInstance().getTableSqlNameFilter()); }
    public String getTableDbName() { return _tableDbName; }
    public String getTableDispName() { return _tableDispName; }
    public String getTablePropertyName() { return _tablePropertyName; }
    public TableSqlName getTableSqlName() { return _tableSqlName; }

    // ===================================================================================
    //                                                                         Column Info
    //                                                                         ===========
    protected final ColumnInfo _columnMultipleFirstId = cci("multiple_first_id", "multiple_first_id", null, null, Integer.class, "multipleFirstId", null, true, false, true, "int4", 10, 0, null, false, null, null, null, null, null, false);
//"int64"
    protected final ColumnInfo _columnMultipleSecondId = cci("multiple_second_id", "multiple_second_id", null, null, Integer.class, "multipleSecondId", null, true, false, true, "int4", 10, 0, null, false, null, null, null, null, null, false);
//"int64"
    protected final ColumnInfo _columnRefFirstId = cci("ref_first_id", "ref_first_id", null, null, Integer.class, "refFirstId", null, false, false, true, "int4", 10, 0, null, false, null, null, "whiteCompoundPk", null, null, false);
//"int64"
    protected final ColumnInfo _columnRefSecondId = cci("ref_second_id", "ref_second_id", null, null, Integer.class, "refSecondId", null, false, false, true, "int4", 10, 0, null, false, null, null, "whiteCompoundPk", null, null, false);
//"int64"

    /**
     * multiple_first_id: {PK, NotNull, int4(10)}
     * @return The information object of specified column. (NotNull)
     */
    public ColumnInfo columnMultipleFirstId() { return _columnMultipleFirstId; }
    /**
     * multiple_second_id: {PK, NotNull, int4(10)}
     * @return The information object of specified column. (NotNull)
     */
    public ColumnInfo columnMultipleSecondId() { return _columnMultipleSecondId; }
    /**
     * ref_first_id: {NotNull, int4(10), FK to white_compound_pk}
     * @return The information object of specified column. (NotNull)
     */
    public ColumnInfo columnRefFirstId() { return _columnRefFirstId; }
    /**
     * ref_second_id: {NotNull, int4(10), FK to white_compound_pk}
     * @return The information object of specified column. (NotNull)
     */
    public ColumnInfo columnRefSecondId() { return _columnRefSecondId; }

    protected List<ColumnInfo> ccil() {
        List<ColumnInfo> ls = newArrayList();
        ls.add(columnMultipleFirstId());
        ls.add(columnMultipleSecondId());
        ls.add(columnRefFirstId());
        ls.add(columnRefSecondId());
        return ls;
    }

    { initializeInformationResource(); }

    // ===================================================================================
    //                                                                         Unique Info
    //                                                                         ===========
    // -----------------------------------------------------
    //                                       Primary Element
    //                                       ---------------
    protected UniqueInfo cpui() {
        List<ColumnInfo> ls = newArrayListSized(4);
        ls.add(columnMultipleFirstId());
        ls.add(columnMultipleSecondId());
        return hpcpui(ls);
    }
    public boolean hasPrimaryKey() { return true; }
    public boolean hasCompoundPrimaryKey() { return true; }

    // ===================================================================================
    //                                                                       Relation Info
    //                                                                       =============
    // cannot cache because it uses related DB meta instance while booting
    // (instead, cached by super's collection)
    // -----------------------------------------------------
    //                                      Foreign Property
    //                                      ----------------
    /**
     * white_compound_pk by my ref_first_id, ref_second_id, named 'whiteCompoundPk'.
     * @return The information object of foreign property. (NotNull)
     */
    public ForeignInfo foreignWhiteCompoundPk() {
        Map<ColumnInfo, ColumnInfo> mp = newLinkedHashMapSized(4);
        mp.put(columnRefFirstId(), WhiteCompoundPkDbm.getInstance().columnPkFirstId());
        mp.put(columnRefSecondId(), WhiteCompoundPkDbm.getInstance().columnPkSecondId());
        return cfi("fk_white_compound_pk_ref", "whiteCompoundPk", this, WhiteCompoundPkDbm.getInstance(), mp, 0, org.dbflute.optional.OptionalEntity.class, false, false, false, false, null, null, false, "whiteCompoundPkRefList", false);
    }

    // -----------------------------------------------------
    //                                     Referrer Property
    //                                     -----------------

    // ===================================================================================
    //                                                                        Various Info
    //                                                                        ============

    // ===================================================================================
    //                                                                           Type Name
    //                                                                           =========
    public String getEntityTypeName() { return "com.mssoftech.dbflute.exentity.WhiteCompoundPkRef"; }
    public String getConditionBeanTypeName() { return "com.mssoftech.dbflute.cbean.WhiteCompoundPkRefCB"; }
    public String getBehaviorTypeName() { return "com.mssoftech.dbflute.exbhv.WhiteCompoundPkRefBhv"; }

    // ===================================================================================
    //                                                                         Object Type
    //                                                                         ===========
    public Class<WhiteCompoundPkRef> getEntityType() { return WhiteCompoundPkRef.class; }

    // ===================================================================================
    //                                                                     Object Instance
    //                                                                     ===============
    public WhiteCompoundPkRef newEntity() { return new WhiteCompoundPkRef(); }

    // ===================================================================================
    //                                                                   Map Communication
    //                                                                   =================
    public void acceptPrimaryKeyMap(Entity et, Map<String, ? extends Object> mp)
    { doAcceptPrimaryKeyMap((WhiteCompoundPkRef)et, mp); }
    public void acceptAllColumnMap(Entity et, Map<String, ? extends Object> mp)
    { doAcceptAllColumnMap((WhiteCompoundPkRef)et, mp); }
    public Map<String, Object> extractPrimaryKeyMap(Entity et) { return doExtractPrimaryKeyMap(et); }
    public Map<String, Object> extractAllColumnMap(Entity et) { return doExtractAllColumnMap(et); }
}

File: src/dbflute/adf/meta/whiteCompoundPkRefdbm.go
package meta

import (
	"github.com/mikeshimura/dbflute/df"
)

type WhiteCompoundPkRefDbm_T struct {
	df.BaseDBMeta
	ColumnMultipleFirstId *df.ColumnInfo
	ColumnMultipleSecondId *df.ColumnInfo
	ColumnRefFirstId *df.ColumnInfo
	ColumnRefSecondId *df.ColumnInfo
}

func (b *WhiteCompoundPkRefDbm_T) GetProjectName() string {
	return df.DBCurrent_I.ProjectName
}

func (b *WhiteCompoundPkRefDbm_T) GetDbCurrent() *df.DBCurrent {
	return df.DBCurrent_I
}

var WhiteCompoundPkRefDbm *WhiteCompoundPkRefDbm_T

func Create_WhiteCompoundPkRefDbm() {
	WhiteCompoundPkRefDbm = new(WhiteCompoundPkRefDbm_T)
	WhiteCompoundPkRefDbm.TableDbName = "white_compound_pk_ref"
	WhiteCompoundPkRefDbm.TableDispName = "white_compound_pk_ref"
	WhiteCompoundPkRefDbm.TablePropertyName = "whiteCompoundPkRef"
	WhiteCompoundPkRefDbm.TableSqlName = new(df.TableSqlName)
	WhiteCompoundPkRefDbm.TableSqlName.TableSqlName = "white_compound_pk_ref"
	WhiteCompoundPkRefDbm.TableSqlName.CorrespondingDbName = WhiteCompoundPkRefDbm.TableDbName

	var whiteCompoundPkRef df.DBMeta
	whiteCompoundPkRef = WhiteCompoundPkRefDbm
	WhiteCompoundPkRefDbm.DBMeta=&whiteCompoundPkRef
	multipleFirstIdSqlName := new(df.ColumnSqlName)
	//colsqlname dayoo multiple_first_id
	multipleFirstIdSqlName.ColumnSqlName = "multiple_first_id"
	multipleFirstIdSqlName.IrregularChar = false
	WhiteCompoundPkRefDbm.ColumnMultipleFirstId = df.CCI(&whiteCompoundPkRef, "multiple_first_id", multipleFirstIdSqlName, "", "", "Integer.class", "multipleFirstId", "", true, false,true, "int4", 10, 0, "",false,"","", "","","",false,"int64")
	multipleSecondIdSqlName := new(df.ColumnSqlName)
	//colsqlname dayoo multiple_second_id
	multipleSecondIdSqlName.ColumnSqlName = "multiple_second_id"
	multipleSecondIdSqlName.IrregularChar = false
	WhiteCompoundPkRefDbm.ColumnMultipleSecondId = df.CCI(&whiteCompoundPkRef, "multiple_second_id", multipleSecondIdSqlName, "", "", "Integer.class", "multipleSecondId", "", true, false,true, "int4", 10, 0, "",false,"","", "","","",false,"int64")
	refFirstIdSqlName := new(df.ColumnSqlName)
	//colsqlname dayoo ref_first_id
	refFirstIdSqlName.ColumnSqlName = "ref_first_id"
	refFirstIdSqlName.IrregularChar = false
	WhiteCompoundPkRefDbm.ColumnRefFirstId = df.CCI(&whiteCompoundPkRef, "ref_first_id", refFirstIdSqlName, "", "", "Integer.class", "refFirstId", "", false, false,true, "int4", 10, 0, "",false,"","", "whiteCompoundPk","","",false,"int64")
	refSecondIdSqlName := new(df.ColumnSqlName)
	//colsqlname dayoo ref_second_id
	refSecondIdSqlName.ColumnSqlName = "ref_second_id"
	refSecondIdSqlName.IrregularChar = false
	WhiteCompoundPkRefDbm.ColumnRefSecondId = df.CCI(&whiteCompoundPkRef, "ref_second_id", refSecondIdSqlName, "", "", "Integer.class", "refSecondId", "", false, false,true, "int4", 10, 0, "",false,"","", "whiteCompoundPk","","",false,"int64")

	WhiteCompoundPkRefDbm.ColumnInfoList = new(df.List)
	WhiteCompoundPkRefDbm.ColumnInfoList.Add(WhiteCompoundPkRefDbm.ColumnMultipleFirstId)
	WhiteCompoundPkRefDbm.ColumnInfoList.Add(WhiteCompoundPkRefDbm.ColumnMultipleSecondId)
	WhiteCompoundPkRefDbm.ColumnInfoList.Add(WhiteCompoundPkRefDbm.ColumnRefFirstId)
	WhiteCompoundPkRefDbm.ColumnInfoList.Add(WhiteCompoundPkRefDbm.ColumnRefSecondId)


	WhiteCompoundPkRefDbm.ColumnInfoMap=make(map[string]int)
	WhiteCompoundPkRefDbm.ColumnInfoMap["multipleFirstId"]=0
		WhiteCompoundPkRefDbm.ColumnInfoMap["multipleSecondId"]=1
		WhiteCompoundPkRefDbm.ColumnInfoMap["refFirstId"]=2
		WhiteCompoundPkRefDbm.ColumnInfoMap["refSecondId"]=3
	    WhiteCompoundPkRefDbm.PrimaryKey = true
    WhiteCompoundPkRefDbm.CompoundPrimaryKey = true
	ui := new(df.UniqueInfo)
	ui.DbMeta = &whiteCompoundPkRef
	ui.Primary = true
	ui.UniqueColumnList = new(df.List)
    ui.UniqueColumnList.Add(WhiteCompoundPkRefDbm.ColumnMultipleFirstId)
    ui.UniqueColumnList.Add(WhiteCompoundPkRefDbm.ColumnMultipleSecondId)

	WhiteCompoundPkRefDbm.PrimaryInfo = new(df.PrimaryInfo)
	WhiteCompoundPkRefDbm.PrimaryInfo.UniqueInfo = ui
	
	var whiteCompoundPkRefMeta df.DBMeta = WhiteCompoundPkRefDbm
	df.DBMetaInstanceHandler_I.TableDbNameInstanceMap["WhiteCompoundPkRef"] = &whiteCompoundPkRefMeta
}
